Implement EVM environmental opcodes

Bhargavasomu opened the pull request:
### What was wrong?
Environment related opcodes weren't implemented.
Related to Issue #254 

### How was it fixed?
Following frontier yellowpaper

#### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://pbs.twimg.com/profile_images/521554275713830913/TBY5IslL.jpeg)

Bhargavasomu commented:
This should be reviewed post merging of #248

lightclient commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+
+def caller(evm: Evm) -> None:
+    """
+    Pushes the address of the caller or transaction sender. The caller can be
```
```suggestion
    Pushes the address of the caller.
```

lightclient commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+def caller(evm: Evm) -> None:
+    """
+    Pushes the address of the caller or transaction sender. The caller can be
+    an EOA or another CA.
```
```suggestion
```

SamWilsn replied:
Off topic, but you can drag to select multiple lines to make a multi-line patch.

lightclient commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+
+def callvalue(evm: Evm) -> None:
+    """
+    Obtain the value included in the transaction responsible for this
```
```suggestion
    Push the value, in wei, sent with the call onto the stack.
```

lightclient commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+    start_index = Uint(pop(evm.stack))
+    calldataload = evm.data[start_index : start_index + 32]
+    # Right pad with 0 so that there are overall 32 bytes.
+    calldataload = calldataload.ljust(32, b"\x00")
```
Aren't the zeroes added here just stripped off next line?

Bhargavasomu replied:
Not really. The zeroes are being padded to the right and lstrip will remove any zeroes in the beginning. I anyways removed this as the `lstrip` is unnecessary.

SamWilsn commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+        # Get the last 20 bytes and consider this as the address
+        address = address[-MAX_ADDRESS_BYTES:]
+
+    account_state = evm.env.state.get(address, None)
```
Hmmmm... Should `state` be a `defaultdict` instead of a `dict`?

SamWilsn replied:
I guess a better way to phrase that question would be: are there cases where `None` and a default `Account` need to be distinguished, or should they be indistinguishable?

Bhargavasomu replied:
I feel using defaultdict can sometimes get dangerous. For example it's common to do things like follow which will grow the size of the state unknowingly and unintentionally.
```
In [1]: from collections import defaultdict

In [2]: ob = defaultdict(int)

In [3]: ob["key1"]
Out[3]: 0

In [4]: ob
Out[4]: defaultdict(int, {'key1': 0})

In [5]: try:
   ...:     ob["key2"]
   ...: except KeyError:
   ...:     pass
   ...: 

In [6]: ob
Out[6]: defaultdict(int, {'key1': 0, 'key2': 0})    # Unintentionally increased the object size.

```

That said, I'm not sure if it is worth it and if it adds to the code complexity (readability wise). Let me know your thoughts. Thanks.

SamWilsn replied:
This doesn't need to be decided now, I was mostly musing about whether this is a good fit for `defaultdict`. Good point about unintentionally increasing the size.

SamWilsn commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+    # Converting start_index to Uint from U256 as start_index + 32 can
+    # overflow U256.
+    start_index = Uint(pop(evm.stack))
+    calldataload = evm.data[start_index : start_index + 32]
```
I would prefer if we avoid shadowing `calldataload` with a different type, though I might be being a tiiiiiny bit pedantic.

Bhargavasomu replied:
Done.

SamWilsn commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+    start_index = Uint(pop(evm.stack))
+    calldataload = evm.data[start_index : start_index + 32]
+    # Right pad with 0 so that there are overall 32 bytes.
+    calldataload = calldataload.ljust(32, b"\x00")
```
I don't think `U256.from_be_bytes` cares if there are less than 32 bytes, so padding should be unnecessary.

Bhargavasomu replied:
The yellowpaper says we should right pad it with 0's instead of left padding, hence changing the value being pushed to the stack.

SamWilsn commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_EXT)
+
+    address = pop(evm.stack).to_be_bytes()
+    if len(address) > MAX_ADDRESS_BYTES:
+        address = address[-MAX_ADDRESS_BYTES:]
+
+    account_state = evm.env.state.get(address, None)
```
Same concern about `len(address) < 20`.

SamWilsn commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+        address = address[-MAX_ADDRESS_BYTES:]
+
+    account_state = evm.env.state.get(address, None)
+    if account_state is None:
+        codesize = U256(0)
+    else:
+        codesize = U256(len(evm.env.state[address].code))
```
Similar question about `defaultdict`.

Bhargavasomu replied:
Responded [here](https://github.com/ethereum/eth1.0-specs/pull/255#discussion_r673042105)

SamWilsn commented on src/ethereum/frontier/vm/instructions/stack.py:
```diff
@@ -81,6 +81,8 @@ def swap_n(evm: Evm, item_number: int) -> None:
     """
     Swap the 1st and Nth items in the stack. All items are 0-indexed from the
     top of the stack.
+    If N is 0, then this is a Noop. Regardless, `swap_n` with N as 0 isn't
+    a valid EVM opcode and hence shouldn't be used.
```
I just wanted to clarify this a bit better, since I was confused between `1st` and `All items are 0-indexed`. Hope that's okay!

```suggestion
    Swap the top and the `item_number` element of the stack, where
    the top of the stack is position zero. If `item_number` is zero, this
    function does nothing (which should not be possible, since there
    is no `SWAP0` instruction.)
```

Bhargavasomu replied:
Makes sense. These changes would be reflecting in #248

SamWilsn commented on src/ethereum/frontier/vm/instructions/bitwise.py:
```diff
@@ -0,0 +1,148 @@
+
+def get_byte(evm: Evm) -> None:
+    """
+    For a word (defined by next top element of the stack), retrieve the
+    Nth byte (0-indexed and defined by top element of stack) from the left
+    to right (MSB to LSB).
```
```suggestion
    For a word (defined by next top element of the stack), retrieve the
    Nth byte (0-indexed and defined by top element of stack) from the left (most significant)
    to right (least significant).
```

Bhargavasomu replied:
Makes sense. These changes would be reflecting in #248

SamWilsn commented on whitelist.txt:
```diff
@@ -58,6 +58,9 @@ vm
 num
 utils
 prepend
+LSB
+MSB
```
```suggestion
```

Bhargavasomu replied:
Makes sense. These changes would be reflecting in #248

Bhargavasomu commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+from ..gas import GAS_BASE, GAS_EXT, GAS_VERY_LOW, subtract_gas
+from ..stack import pop, push
+
+# Address can be only 20 bytes
+MAX_ADDRESS_BYTES = 20
```
For now, this is the only place which this constant is being used. In the future if there are similar usages, I will move this to `eth_types.py`. Let me know how this sounds.

SamWilsn replied:
I think the way you have it is fine for now, especially since we need to rework the array types anyway.

Bhargavasomu commented on src/ethereum/frontier/vm/instructions/comparison.py:
```diff
@@ -0,0 +1,172 @@
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+
+    x = pop(evm.stack)
+    y = pop(evm.stack)
```
Makes sense. These changes would be reflecting in #248

Bhargavasomu commented on src/ethereum/frontier/vm/instructions/comparison.py:
```diff
@@ -0,0 +1,172 @@
+from ..stack import pop, push
+
+
+def lt(evm: Evm) -> None:
```
Ok, will expand them. These changes would be reflecting in #248

Bhargavasomu commented on src/ethereum/frontier/vm/instructions/computation.py:
```diff
@@ -1,6 +1,6 @@
 """
```
The name has been changed to control_flow.py. This branch has not been rebased on the latest master. Though #248 is rebased on master, and you will find the file name changed to `control_flow.py` over there. Resolving this convo.

Bhargavasomu commented:
Thanks for your review @lightclient @SamWilsn . I'll be addressing these comments post merging of #248, as this PR also involves the commit mentioned in #248 . Thank you.

Bhargavasomu commented on src/ethereum/frontier/vm/instructions/environment.py:
```diff
@@ -0,0 +1,244 @@
+    # custom test cases.
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_EXT)
+
+    address = pop(evm.stack).to_be_bytes()
+    if len(address) > MAX_ADDRESS_BYTES:
+        # Get the last 20 bytes and consider this as the address
+        address = address[-MAX_ADDRESS_BYTES:]
```
Thanks for catching this. This is an issue. For some reason I assumed that `dict[b'\x00\x01']` and `dict[b'\x01']` would give the same result. I have changed this.

Bhargavasomu commented:
@lightclient I believe I have addressed all your comments mentioned earlier. That said, given you would be busy in ethcc this week, we'll be merging this PR. Please do take a look later when free. Thanks!

@SamWilsn can you please merge this PR? Thanks.
